### PR TITLE
Add DeepSeek model client

### PR DESCRIPTION
**Pull request by mxsl-gr**
Hi, this PR is add `DeepSeek` model client and has passed unit testing.
I can provide my api_key if needed for testing

the  PR content:

- DeepSeek chat client
- spring starter
- unit test
- chat client documents

For some reasons, products from OpenAI and others can't be directly used in Chinese Mainland.

`DeepSeek` is a strong, economical, and efficient Mixture of Experts (MoE) language model, with an API pricing of $0.14/$0.28 per 1 million tokens.

the link: [DeepSeek](https://www.deepseek.com/en)

If they can supported, it will further aid spring-ai to promotion in Chinese Mainland.

If necessary, I can take care of subsequent maintenance since I'm currently using them.

The `Moonshot` PR is #596
The `ZhiPuAI` PR is #623
The `MiniMax` PR is #628

### CODE REVIEW

**Comment by mxsl-gr** [comment]
> There are a lot of AI companies, and a lot of AI models, so the code can't be finished by creating clients all the time. Another thing is that deepseek has already said in the document that his API is fully compatible with openAI, that is to say, chatclient set the proxy address can call deepseek smoothly, i think the project maintainer needs to consider this problem: not to add a new model, but to add a new way to call，above are just my own thoughts。

hi, I agree with your point. I have two thoughts.
﻿
if the companies provided an API standard compatible with OpenAI, in most cases, you can use it by just replacing the base url and API key, it's great.
﻿
while the OpenAI API design is not a standard, there are indeed many followers of its design, especially in mainland China. But this need the followers maintaining compatibility with the OpenAI API design. During the stage when their features are converging, it is predictable as they want to lower the barrier for potential users. But if they continue to develop, their paths may not necessarily converge.
A somewhat inappropriate example is `Quarkus` and `Spring`. `Quarkus` is relatively aggressive in native aspects, but they need to lower the barrier for developers using `Spring`. So, they provide `Spring` extensions to work in some cases to achieve low-barrier migration. However, if `Spring` API gets new designs, `Quarkus` may not necessarily follow them continuously.
﻿
Secondly, the two are not completely compatible. I encountered this with the `n` parameter. I have a scenario where I need to generate multiple options and then score them, but Deepseek currently does not support the `n` parameter. I think that if the API does not support it, it’s best not to include it in the configuration. This is something that cannot be achieved if using `OpenAI` implementation.
﻿
in conclusion, I think widely used model providers can still have independent implementations, but it doesn’t hurt to add a new way to call.

**Comment by mxsl-gr** [comment]
this PR has completed the refactoring, has been squashed and force pushed again.

**Comment by markpollack** [comment]
In my experience at the model level, I agree that once you get into the details, "api compatability" is surface level.  On the other hand, there is indeed a model "zoo" that can add to maintenance and carries risk of them disappearing.

I suspect that over time, once the market decides on winners and losers in the marketplace, there will be consolidation and some models in spring ai will go away.  As a concrete example of that, Google's [PAaM API is deprecated](https://ai.google.dev/palm_docs/deprecation) so we will be removing it (I'll create an issue).

**Comment by markpollack** [comment]
I'm still not sure what to do here.  @mxsl-gr have you tried to use Spring AI's OpenAI support to talk with this model?

**Comment by mxsl-gr** [comment]
> I'm still not sure what to do here. Other opinions? @mxsl-gr have you tried to use Spring AI's OpenAI support to talk with this model?

hi @markpollack , i have been busy with other work recently. we are using a customized version built on my local branch, which merges this branch with other strange features. for example, we have a multi-level model that can handle tasks at different levels to improve speed or reduce costs, as well as a client pool and polling mechanisms to enhance throughput and redundancy.
﻿i just ran the `OpenAI` model unit tests using `DeepSeek`'s base url and key, and they worked for most scenarios, i think with configuration the `OpenAI` model can be used in most cases. but there are some properties not supported by `DeepSeek`, such as the parameter `n` mentioned earlier. if we ignore those properties, so this PR can be closed.
﻿
however, considering that DeepSeek has a significant user base in Mainland China, now even surpassing `ZhiPuAI` and `Minimax`. i think having an independent implementation isn't necessarily a bad idea. if that's the case, i might need to adjust this branch, including the implementation of function calls and adding more unit tests.

**Comment by markpollack** [comment]
Thanks for the update.  I did take a look at the github project, it doesn't seem active - https://github.com/deepseek-ai/DeepSeek-V2/graphs/contributors  what is your opinion.  Does function calling with OpenAI 'compatability' work.  In practice, for chat models, I haven't seen much use of 'n' in the wild, say as compared to image models.

**Comment by markpollack** [comment]
oops, didn't mean to close.

**Comment by markpollack** [comment]
Also, llama3 has a similar commit history - minimal - I guess all the work is up front!  https://github.com/meta-llama/llama3/graphs/contributors

**Comment by mxsl-gr** [comment]
> Hi, everyone! Do you have any plans for when the branch will be merged into the main branch?

hi, for some reason, this PR hasn’t been merged yet.
but I think now is a good timing, our customized version on local branch is still being actively updated. but certain features, like `Prefix Completion`, are difficult to implement in a non-intrusive way within the current `Spring AI` project code, I’ll try to simplify and adjust it later.
I also hope it gets merged since DeepSeek is the model I use the most.

**Comment by mxsl-gr** [comment]
hi, @markpollack 

this PR has completed the refactoring, implemented beta features like `prefix completion`.

the Function Call feature was added at this commit, but is not very stable, because the current version  (V2.5 ~ V3) of DeepSeek's modal Function Call is unstable, the tests related to the function call have been temporarily disabled. 

All other tests have passed.

References: [DeepSeek Function Call](https://api-docs.deepseek.com/guides/function_calling)

**Comment by markpollack** [comment]
I’m so very sorry for not being timely.  My bad.

Let’s find a time to zoom. I think there is a mixture of approaches, using OpenAI comparability but also allowing for a more dedicated model.  This can be something we develop in the spring-ai-community github.  With the GA nearing we need to settle on a supported set of models (we need to support them for commercial support contexts for 3 years or more).  A dedicated model can evolve outside out GA timeline and later be incorporated or live side by side as with spring data and and spring cloud community projects.  

Please email me for a good time to chat.  
again , sincere apologies for not relying sooner.

**Comment by mxsl-gr** [comment]
> I’m so very sorry for not being timely. My bad.
> 
> Let’s find a time to zoom. I think there is a mixture of approaches, using OpenAI comparability but also allowing for a more dedicated model. This can be something we develop in the spring-ai-community github org. With the GA nearing we need to settle on a supported set of models (we need to support them for commercial support contexts for 3 years or more). A dedicated model can evolve outside out GA timeline and later be incorporated or live side by side as with spring data and and spring cloud community projects. You can commit directly and be in charge without me or anyone else on the team being a blocking factor.
> 
> Please email me for a good time to chat. again , sincere apologies for not relying sooner.

hi, @markpollack 
﻿
i'm sorry for the delayed response, i just got back from a long trip.
﻿
i understand the challenge of maintaining multiple models and that you’re suggesting we manage some of them through community projects, similar to how `spring-data-elasticsearch` relates to `spring-data`. on that point, I’m fine with the idea.

**Comment by eternachen** [comment]
Hi, @mxsl-gr @markpollack is there any plan to merge this PR?

**Comment by mxsl-gr** [comment]
hi, @markpollack 

The function call changes and CoT support for `deepseek-reasoner` have been completed, and all unit tests have passed.

**Comment by markpollack** [comment]
The PR is basically impossible to rebase as so much has changed.  Also note that autoconfig now has a different organization.  

I've extracted the files from the PR and merged the core classes into a new module `spring-ai-deepseek`.

See a5920a17fa4b9c0d28d33f879e2b61386fb2b6eb

Please can you make a new PR with the remaining parts from this PR that need to be added?

**Comment by mxsl-gr** [comment]
> The PR is basically impossible to rebase as so much has changed. Also note that autoconfig now has a different organization.
> 
> I've extracted the files from the PR and merged the core classes into a new module `spring-ai-deepseek`.
> 
> See [a5920a1](https://github.com/spring-projects/spring-ai/commit/a5920a17fa4b9c0d28d33f879e2b61386fb2b6eb)
> 
> Please can you make a new PR with the remaining parts from this PR that need to be added?

Sorry, I missed the changes in the project structure. I’ll take care of it today.
